File: agrotech-back/src/main/java/br/com/agrotech/agrotech_api/service/FazendaService.java
package br.com.agrotech.agrotech_api.service;

import br.com.agrotech.agrotech_api.model.Fazenda;
import br.com.agrotech.agrotech_api.repository.FazendaRepository;
import org.springframework.beans.factory.annotation.Autowired;
import org.springframework.stereotype.Service;

import java.util.List;
import java.util.Optional;

/**
 * Classe de serviço responsável por gerenciar as regras de negócio da entidade {@link Fazenda}.
 * Intermedeia a comunicação entre a camada de controle (Web) e o repositório de dados.
 */
@Service
public class FazendaService {

    @Autowired
    private FazendaRepository fazendaRepository;

    /**
     * Recupera todas as propriedades rurais cadastradas no sistema.
     *
     * @return {@link List} contendo todas as instâncias de {@link Fazenda}.
     */
    public List<Fazenda> buscarTodas() {
        return fazendaRepository.findAll();
    }

    /**
     * Busca uma fazenda específica pelo seu identificador único.
     *
     * @param id Identificador único da fazenda.
     * @return Um {@link Optional} contendo a {@link Fazenda} caso encontrada.
     */
    public Optional<Fazenda> buscarPorId(Long id) {
        return fazendaRepository.findById(id);
    }

    /**
     * Realiza o cadastro ou atualização de uma propriedade rural no sistema.
     *
     * @param fazenda Objeto contendo os dados da fazenda a ser persistida.
     * @return A instância de {@link Fazenda} salva com seu ID gerado.
     */
    public Fazenda salvar(Fazenda fazenda) {
        return fazendaRepository.save(fazenda);
    }

    /**
     * Atualiza os dados de uma fazenda existente no sistema.
     *
     * @param id Identificador da fazenda a ser atualizada.
     * @param fazendaAtualizada Objeto contendo os novos dados da fazenda.
     * @return A instância de {@link Fazenda} atualizada.
     * @throws RuntimeException Se a fazenda não for encontrada pelo ID.
     */
    public Fazenda atualizar(Long id, Fazenda fazendaAtualizada) {
        return fazendaRepository.findById(id)
                .map(fazenda -> {
                    fazenda.setNome(fazendaAtualizada.getNome());
                    fazenda.setLocalizacao(fazendaAtualizada.getLocalizacao());
                    fazenda.setAreaHectares(fazendaAtualizada.getAreaHectares());
                    return fazendaRepository.save(fazenda);
                })
                .orElseThrow(() -> new RuntimeException("Fazenda não encontrada com o ID: " + id));
    }

    /**
     * Remove uma propriedade rural do sistema pelo seu identificador.
     *
     * @param id Identificador da fazenda a ser removida.
     * @throws RuntimeException Se a fazenda não for encontrada.
     */
    public void deletar(Long id) {
        if (!fazendaRepository.existsById(id)) {
            throw new RuntimeException("Fazenda não encontrada com o ID: " + id);
        }
        fazendaRepository.deleteById(id);
    }
}

File: agrotech-front/src/services/FazendaService.ts
import { api } from './api';
import type { Fazenda } from '../types/Fazenda';

/**
 * Serviço responsável por integrar a camada de interface web com a API Spring Boot
 * referente ao gerenciamento das propriedades rurais (Fazendas).
 */
export const fazendaService = {
    /**
     * Recupera a lista completa de fazendas cadastradas no sistema.
     *
     * @returns {Promise<Fazenda[]>} Uma promessa contendo a lista de fazendas.
     */
    listarTodas: async (): Promise<Fazenda[]> => {
        const response = await api.get<Fazenda[]>('/fazendas');
        return response.data;
    },

    /**
     * Busca os detalhes de uma fazenda específica pelo seu identificador.
     *
     * @param {number} id - Identificador único da fazenda.
     * @returns {Promise<Fazenda>} Uma promessa contendo a instância da fazenda encontrada.
     */
    buscarPorId: async (id: number): Promise<Fazenda> => {
        const response = await api.get<Fazenda>(`/fazendas/${id}`);
        return response.data;
    },

    /**
     * Realiza o cadastro de uma nova propriedade rural no sistema.
     *
     * @param {Fazenda} fazenda - Objeto contendo os dados da nova fazenda.
     * @returns {Promise<Fazenda>} Uma promessa contendo a fazenda cadastrada e com ID gerado.
     */
    cadastrar: async (fazenda: Fazenda): Promise<Fazenda> => {
        const response = await api.post<Fazenda>('/fazendas', fazenda);
        return response.data;
    },

    /**
     * Atualiza as informações de uma fazenda já existente.
     *
     * @param {number} id - Identificador único da fazenda a ser alterada.
     * @param {Fazenda} fazenda - Objeto com os novos dados atualizados.
     * @returns {Promise<Fazenda>} Uma promessa contendo a fazenda atualizada.
     */
    atualizar: async (id: number, fazenda: Fazenda): Promise<Fazenda> => {
        const response = await api.put<Fazenda>(`/fazendas/${id}`, fazenda);
        return response.data;
    },

    /**
     * Remove um registro de fazenda do sistema através do seu ID.
     *
     * @param {number} id - Identificador único da fazenda a ser excluída.
     * @returns {Promise<void>} Promessa resolvida após a exclusão com sucesso.
     */
    deletar: async (id: number): Promise<void> => {
        await api.delete(`/fazendas/${id}`);
    },
};